### DISCUSSION TITLE
Roda & WebSockets

### DISCUSSION THREAD

**Discussion by hmdne** [jeremyevans/roda#230] [General]
What do you use for WebSocket support with Roda? I know it's a hard question, because most WebSocket libraries in Ruby are using EventMachine which is basically unmaintained (last release - 2018).

I found this thing, but I couldn't make it work (2019), also it depends on Falcon being used and I understand it doesn't have to be the case.
* https://github.com/socketry/roda-websockets

ActionCable depends on the following gems:
* https://github.com/socketry/nio4r
* https://github.com/faye/websocket-driver-ruby

Maybe it would be a good idea to replicate what ActionCable is doing?

Anyway, the main reason why I start this discussion is to find the best ideas to go forward.

**Comment by jeremyevans** [comment]
Roda 2 had a websockets plugin, which should still work in Roda 3: https://github.com/jeremyevans/roda/blob/2.29.0/lib/roda/plugins/websockets.rb

It works, but the tests were flaky and depended on thin, so I removed it in Roda 3.  If you would like websockets support in Roda and don't want to use something that depends on Falcon, I recommend copying the Roda 2 websockets plugin and modifying it as you see fit.

**Comment by hmdne** [comment]
Good to know it was here before, but also as I mentioned above, it would be great to find or create something without EventMachine and Thin dependency.

**Comment by adam12** [comment]
I'm using Roda and Websockets, but only indirectly. I use the [Iodine](https://github.com/boazsegev/iodine) app server, and when there's a specific request for websockets, I return an upgrade response with an Iodine handler. It works well.

If I don't have bi-directional communication requirements, I use [MessageBus](https://github.com/jeremyevans/roda-message_bus).
